### PR TITLE
VAnchor proof generation for `wasm-utils`,`sdk-core`

### PR DESCRIPTION
**Pull request by AhmedKorim**
# Overview
After this PR webb.js will support generating ZKP for vanchor API 

# Checklist
- [x] Proof input for vanchor
- [x] implemente `/proof/vanchor.rs` `create_proof`
- [x] Rust/wasm tests
- [x] Update sdk-core
- [x] TS Unit tests for proof generation
- [ ] <del> Integration tests for VAnchor transfare</del>

### CODE REVIEW

**Comment by AhmedKorim** [comment]
@drewstone  This is ready but still one piece missing in filling the gaps between the inputs `(2,16)`
Input UTXOs should have an index what value to use for that?

File: packages/api-providers/src/polkadot/mixer-withdraw.ts [hunk 1]
```diff
@@ -167,27 +181,22 @@ export class PolkadotMixerWithdraw extends MixerWithdraw<WebbPolkadot> {
...
-        // Fetch the internal ID for the intended withdraw chain
-        const internalId = chainTypeIdToInternalId(parseChainIdType(Number(noteParsed.note.targetChainId)));
-
         const relayerWithdrawPayload = relayerMixerTx.generateWithdrawRequest(
           {
             baseOn: 'substrate',
             contractAddress: '',
             endpoint: '',
             // TODO change this from the config
-            name: chainIdToRelayerName(internalId)
```
Comment (nepoche) [inline packages/api-providers/src/polkadot/mixer-withdraw.ts:186]: Why is this changed to 'localnode'?

File: packages/sdk-core/src/proving/proving-manager-thread.ts [hunk 2]
```diff
@@ -98,38 +124,94 @@ export class ProvingManagerWrapper {
...
+
+    if (protocol === 'mixer' || protocol === 'anchor') {
+      const proof = proofOutput.proof;
+
+      const proofPayload: ProofI<'anchor'> = {
+        nullifierHash: proof.nullifierHash,
+        proof: proof.proof,
+        root: proof.root,
+        roots: proof.roots
+      };
```
Comment (nepoche) [inline packages/sdk-core/src/proving/proving-manager-thread.ts:200]: Should the mixer return a different `ProofI<>`?

Comment (AhmedKorim) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: mixer/anchor shares the same proof type, but here it should be `mixer` for sure just for being semantic

Comment (drewstone) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: We should separate this as mixer doesn't have roots object.

Comment (drewstone) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: @AhmedKorim do you agree?

Comment (AhmedKorim) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: This is a wapper around what `arkworks-gadgets` exposes
```rust

pub struct MixerProof {
	pub proof: Vec<u8>,
	pub leaf_raw: Vec<u8>,
	pub nullifier_hash_raw: Vec<u8>,
	pub root_raw: Vec<u8>,
	pub public_inputs_raw: Vec<Vec<u8>>,
}

```

we can do that update after doing it in  `arkworks-gadgets`

Comment (AhmedKorim) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: https://github.com/webb-tools/arkworks-gadgets/blob/a2c494c30d3ee31610f1eb6576c8e3acd3c0cf1a/arkworks-setups/src/common.rs#L54-L54

Comment (AhmedKorim) [reply packages/sdk-core/src/proving/proving-manager-thread.ts:225]: you are right guys

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  const note = noteBuilder.build();
+
+  if (index !== undefined) {
+    note.mutateIndex(String(index));
+  }
+
+  return note;
+}
+
+const vanchorBn2542_2_2 = setupKeys('vanchor', 'Bn254', 2, 2, 2);
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:46]: Can we add another underscore here? `vanchorBn254_2_2_2` and `vanchorBn254_2_16_2`

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+      leavesMap: leavesMap,
+      outputConfigs: [outputConfig1, outputConfig2],
+      provingKey: keys.pk,
+      publicAmount: String(publicAmount),
+      roots: rootsSet
+
+    };
+
+    const data = await provingManager.proof('vanchor', setup);
+
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:92]: Extra spaces unnecessary.

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    const isValidProof = verify_js_proof(data.proof, data.publicInputs, u8aToHex(keys.vk).replace('0x', ''), 'Bn254');
+
+    expect(isValidProof).to.deep.equal(true);
+  });
+
+  it('should proof js for VAnchor with 16 inputs and 16 indices', async () => {
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:192]:
```suggestion
  it('should prove using WASM API for VAnchor with 16 inputs and 16 indices', async () => {
```

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+    };
+    const data = await provingManager.proof('vanchor', setup);
+
+    const isValidProof = verify_js_proof(data.proof, data.publicInputs, u8aToHex(keys.vk).replace('0x', ''), 'Bn254');
+
+    expect(isValidProof).to.deep.equal(true);
+  });
+
+  it('should proof js for VAnchor with three inputs amd three indices', async () => {
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:146]:
```suggestion
  it('should prove using WASM API for VAnchor with three inputs amd three indices', async () => {
```

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    };
+
+    const data = await provingManager.proof('vanchor', setup);
+
+    const isValidProof = verify_js_proof(data.proof, data.publicInputs, u8aToHex(keys.vk).replace('0x', ''), 'Bn254');
+
+    expect(isValidProof).to.deep.equal(true);
+  });
+
+  it('should proof js for VAnchor with two inputs and two indices', async () => {
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:98]:
```suggestion
  it('should prove using WASM API for VAnchor with two inputs and two indices', async () => {
```

File: packages/sdk-core/src/__test__/proving.spec.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    };
+
+    const data = await provingManager.proof('vanchor', setup);
+
+    const isValidProof = verify_js_proof(data.proof, data.publicInputs, u8aToHex(keys.vk).replace('0x', ''), 'Bn254');
+
+    expect(isValidProof).to.deep.equal(true);
+  });
+
+  it('should fail to proof  for VAnchor with 16 inputs and 16 indices with invalid amounts', async () => {
```
Comment (drewstone) [inline packages/sdk-core/src/__test__/proving.spec.ts:238]:
```suggestion
  it('should fail to prove for VAnchor with 16 inputs and 16 indices with invalid amounts', async () => {
```

File: packages/sdk-core/src/note.ts [hunk 9]
```diff
@@ -72,6 +72,18 @@ export class Note {
     }
   }
 
+  /**
```
Comment (drewstone) [inline packages/sdk-core/src/note.ts:75]: This comment is not correct. There's no param. Please fix.

File: packages/sdk-core/src/proving/proving-manager-thread.ts [hunk 10]
```diff
@@ -32,16 +30,43 @@ export type ProvingManagerSetupInput = {
...
+type PMEvents<T extends NoteProtocol = 'mixer'> = {
+  proof: [T, ProvingManagerSetupInput<T>];
   destroy: undefined;
 };
 
+/**
+ * @param roots - Roots for anchor API
+ * @param refreshCommitment - Refresh commitment in hex representation ( without prefix `0x` ) Required for anchor, ignored for the mixer
+ * */
+export type AnchorPMSetupInput = MixerPMSetupInput & {
```
Comment (drewstone) [inline packages/sdk-core/src/proving/proving-manager-thread.ts:46]: Can we just keep them separate from the Mixer? Especially since this looks like it probably takes in other parameters we don't even need, right? Just keep them all well-defined and separate so it's clear.

File: packages/sdk-core/src/proving/proving-manager-thread.ts [hunk 11]
```diff
@@ -57,8 +82,8 @@ export class ProvingManagerWrapper {
 
         switch (key) {
           case 'proof': {
-            const input = message.proof!;
-            const proof = await this.prove(input);
+            const [proto, input] = message.proof!;
```
Comment (drewstone) [inline packages/sdk-core/src/proving/proving-manager-thread.ts:85]: Let's use `protocol`, better to be verbose here please.

File: packages/sdk-core/src/proving/proving-manager.ts [hunk 12]
```diff
@@ -1,10 +1,23 @@
...
 // SPDX-License-Identifier: Apache-2.0
 
+import type { NoteProtocol } from '@webb-tools/wasm-utils';
+
 import { ProvingManagerSetupInput, ProvingManagerWrapper } from '@webb-tools/sdk-core/proving/proving-manager-thread.js';
-import { Proof } from '@webb-tools/wasm-utils';
 
-export type ProofI = Omit<Proof, 'free'>;
+type VAnchorProof = {
+  readonly inputUtxos: Array<any>;
```
Comment (drewstone) [inline packages/sdk-core/src/proving/proving-manager.ts:9]: Is there no way to type these arrays? Can we not use `Utxo` type?

File: packages/wasm-utils/src/note/mod.rs [hunk 13]
```diff
@@ -766,6 +815,37 @@ impl JsNote {
...
+	#[wasm_bindgen(js_name = defaultUtxoNote)]
+	pub fn default_utxo_note(&self) -> Result<JsNote, OperationError> {
+		let mut new_note = self.clone();
+		let chain_id: u64 = self
+			.target_chain_id
+			.parse()
+			.map_err(|_| OpStatusCode::InvalidTargetChain)?;
+
+		let utxo = vanchor::generate_secrets(
+			0,
```
Comment (drewstone) [inline packages/wasm-utils/src/note/mod.rs:827]: Is this meant to be the chain_id? Recall that a default UTXO still needs to have the chain ID where it is being "spent". So this function will need to have access to the chain ID where the transaction is taking place.

**Review by drewstone — CHANGES_REQUESTED** [review]
Initially looks good but do want to fix up these issues. Will do a pass over tests once addressed.

File: packages/wasm-utils/src/note/mod.rs (hunk 13 above)
Comment (AhmedKorim) [reply packages/wasm-utils/src/note/mod.rs:827]: Yeah this is using the `traget_chain_id` value for the` chain_id` of the empty UTXO

File: packages/sdk-core/src/proving/proving-manager.ts (hunk 12 above)
Comment (AhmedKorim) [reply packages/sdk-core/src/proving/proving-manager.ts:9]: Handled types here

File: packages/wasm-utils/src/note/mod.rs (hunk 13 above)
Comment (AhmedKorim) [reply packages/wasm-utils/src/note/mod.rs:827]: This is consuming the `target_chain_id` and using it as the value of the default UTXO `chain_id`
As for input, all UTXOs share the same `chain_id`
